kernel/arch/dreamcast/hardware/spu.c

```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
     }
 }
 
+void spu_memload_sq(uint32 dst, void *src_void, int length) {
+    uint8 *src = (uint8 *)src_void;
+
+    /* Make sure it's an even number of 32-bit words and convert the
+       count to a 32-bit word count */
+    if(length & 3) {
+        length = (length + 4) & ~3;
+    }
+
+    /* Add in the SPU RAM base (cached area) */
+    dst += 0x00800000;
+
+    /* Using SQs for all that is divisible by 32 */
+    const int aligned_len = length & ~31;
+    length &= 31;
+
+    int old = irq_disable();
+    do { } while(*(vuint32 *)0xa05f688c & (1 << 5)) ; // FIFO_SH4
+    do { } while(*(vuint32 *)0xa05f688c & (1 << 4)) ; // FIFO_G2
+    sq_cpy((void *)dst, src, aligned_len);
+    irq_restore(old);
+
+    if(length > 0) {
+        /* Non-cached area */
+        dst += 0xa0000000;
```

Comment:
This should almost certainly be an `|=` not a `+=`.

```suggestion
        /* Make sure the destination is in a non-cached area */
        dst |= 0xa0000000;
```

Reply:
And maybe using the nice new un-magic'd MEM_AREA_P2_BASE?

Reply:
Good point. Updated to use the macro and another suggestion added to include the appropriate header file (even though it is indirectly included through sq.h).